src/gc/gc.d

```diff
@@ -2803,19 +2799,36 @@ struct Pool
         }
     }
 
-    void clrBitsSmallSweep(size_t dataIndex, GCBits.wordtype toClear) nothrow
+    void freePageBits(size_t pagenum, in ref PAGEBITS toFree) nothrow
     {
-        immutable toKeep = ~toClear;
+        import core.internal.traits : staticIota;
+
+        immutable beg = pagenum * (PAGESIZE / 16 / GCBits.BITS_PER_WORD);
+
+        foreach (i; staticIota!(0, PAGEBITS.length))
+            freebits.data[beg + i] |= toFree[i];
+
         if (finals.nbits)
-            finals.data[dataIndex] &= toKeep;
+        {
+            foreach (i; staticIota!(0, PAGEBITS.length))
+                finals.data[beg + i] &= ~toFree[i];
+        }
         if (structFinals.nbits)
-            structFinals.data[dataIndex] &= toKeep;
+        {
+            foreach (i; staticIota!(0, PAGEBITS.length))
+                structFinals.data[beg + i] &= ~toFree[i];
+        }
 
-        noscan.data[dataIndex] &= toKeep;
-        appendable.data[dataIndex] &= toKeep;
+        foreach (i; staticIota!(0, PAGEBITS.length))
+            noscan.data[beg + i] &= ~toFree[i];
+        foreach (i; staticIota!(0, PAGEBITS.length))
+            appendable.data[beg + i] &= ~toFree[i];
```

Comment:
BTW, I always though dmd spends a lot of time in CSE. Lesson learnt never trust any of it's optimizations, even for such basic things as `beg + i`. Not to mention the register allocator.